# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

### Prediction:

1. Veridict
    * My prediction is that Random Forests Classifier will perform better. 

2. Rationale
    * The dataset has lots of features (57 in total)
    * Logistic Regression is a linear regression and will probably tend to overfit
    * Random Forest is robust against overfitting
    * Robust for non linear data
    * Runs efficiently on large datasets
    * And allows feature selection



## Split the Data into Training and Testing Sets

In [5]:
# Create the labels set `y` and features DataFrame `X`

y = data['spam']
X = data.drop(columns=['spam'])

In [6]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.

y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [7]:
# Split the data into X_train, X_test, y_train, y_test

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)
X_train.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
1173            0.00                0.0           0.00           0.0   
1954            0.67                0.0           0.00           0.0   
256             0.00                0.0           1.42           0.0   
3341            0.00                0.0           0.24           0.0   
4469            0.00                0.0           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
1173           0.00            0.00              1.23                0.00   
1954           0.00            0.00              0.00                0.00   
256            0.71            0.00              0.00                0.71   
3341           0.09            0.04              0.00                0.00   
4469           0.00            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
1173             0.00            0.00  ...                   0.0        0.000   
1954             0.00            0.00  ...                   0.0        0.000   
256              0.00            0.71  ...                   0.0        0.000   
3341             0.04            0.00  ...                   0.0        0.014   
4469             0.00            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
1173        0.000          0.0        1.939        0.000        0.000   
1954        0.055          0.0        0.000        0.000        0.000   
256         0.000          0.0        0.931        0.000        0.000   
3341        0.148          0.0        0.014        0.044        0.007   
4469        0.000          0.0        0.000        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
1173                       1.379                           8   
1954                       1.644                          13   
256                       12.055                         117   
3341                       2.112                          26   
4469                       1.500                           2   

      capital_run_length_total  
1173                        40  
1954                        74  
256                        217  
3341                      1223  
4469                         3  

[5 rows x 57 columns]

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [8]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance

scaler = StandardScaler()

In [9]:
# Fit the Standard Scaler with the training data

X_scaler = scaler.fit(X_train)

In [11]:
# Scale the training data

X_train_scaled = X_scaler.transform(X_train) 
X_test_scaled = X_scaler.transform(X_test)   

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. Use a `random_state` of `1`. 

In [12]:
# Create and fit a Logistic Regression model
from sklearn.linear_model import LogisticRegression

lr_model = LogisticRegression(random_state=1)
lr_model.fit(X_train_scaled, y_train)

LogisticRegression(random_state=1)

In [13]:
# Make and save testing predictions with the trained Logistic Regression model using the test data

lr_predictions = lr_model.predict(X_test_scaled)

# Review the predictions

lr_predictions

array([0, 0, 0, ..., 1, 0, 0], dtype=int64)

In [25]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.

from sklearn.metrics import classification_report

accurancy_score = accuracy_score(y_test, lr_predictions)

train_score = lr_model.score(X_train_scaled, y_train)
test_score = lr_model.score(X_test_scaled, y_test)

print(f"Accuracy Score: {accurancy_score:.3f}")
print(f"Training Score: {train_score:.3f}")
print(f"Testing Score: {test_score:.3f}")

# Generate the classification report
class_report = classification_report(y_test, lr_predictions)
print("\nClassification Report:")
print(class_report)


Accuracy Score: 0.923
Training Score: 0.926
Testing Score: 0.923

Classification Report:
              precision    recall  f1-score   support

           0       0.91      0.96      0.94       676
           1       0.94      0.87      0.90       475

    accuracy                           0.92      1151
   macro avg       0.93      0.91      0.92      1151
weighted avg       0.92      0.92      0.92      1151



## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. Use a `random_state` of `1`. 

In [15]:
# Create and fit a Random Forest Classifier model
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(random_state=1)
rf_model.fit(X_train_scaled, y_train)

RandomForestClassifier(random_state=1)

In [16]:
# Make and save testing predictions with the trained Random Forest Classifier model using the test data

rf_predictions = rf_model.predict(X_test_scaled)

# Review the predictions

rf_predictions

array([0, 0, 0, ..., 1, 0, 0], dtype=int64)

In [ ]:
from sklearn.metrics import classification_report

# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.

accurancy_score = accuracy_score(y_test, rf_predictions)

train_score = rf_model.score(X_train_scaled, y_train)
test_score = rf_model.score(X_test_scaled, y_test)

print(f"Accuracy Score: {accurancy_score:.3f}")
print(f"Training Score: {train_score:.3f}")
print(f"Testing Score: {test_score:.3f}")

# Generate the classification report
class_report = classification_report(y_test, rf_predictions)
print("\nClassification Report:")
print(class_report)


Accuracy Score: 0.958
Training Score: 0.999
Testing Score: 0.958

Classification Report:
              precision    recall  f1-score   support

           0       0.95      0.98      0.97       676
           1       0.98      0.92      0.95       475

    accuracy                           0.96      1151
   macro avg       0.96      0.95      0.96      1151
weighted avg       0.96      0.96      0.96      1151



## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

* Random Forest Classifier performed better.
* This aligns to my prediction
* The Logistic Regression presented an accuracy of 92.3% while Random Forest model presented an accuracy of 95.8%
* Looking into more detail with the classification report, we observe a higher F1-Score for the Random Forest

